runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -118,14 +120,16 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
             try {
               for (;;) {
                 Optional<Boolean> result = checkForSuccess(job);
-                if (result.isPresent()) {
+                if (result.isPresent() && (!result.get() || atMaxWatermark(job))) {
```

Comment:
Clarify what checkForSuccess() and atMaxWatermark() do in its javadoc
I think you need to update checkForSuccess to make both have clear meanings.

For example, "Also it's passible that the streaming pipeline doesn't use PAssert. So checkForSuccess() will return true before job is finished. "
This means the job might fail, even after checkForSuccess returns true. That doesn't sounds right.